### PR TITLE
NotImplemented as a local exception

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

NotImplementedException from sun.reflect.generics.reflectiveObjects is not standard/guarateed by the jvm so a custom implementation is in order.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java [hunk 1]
```diff
@@ -72,7 +70,7 @@ public static PreallocationType getPreallocationType(final Storage.ProvisioningT
             case FAT:
                 return PreallocationType.Full;
             default:
-                throw new NotImplementedException();
+                throw new NotImplementedProvisioningTypeException();
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: What if instead, you throw a cloud runtime with a message describing the problem/error here?

Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: That is certainly an option, @rafaelweingartner . I had discussed with @nvazquez and he suggested the commons-lang NotImplementedException as it is loaded through the storage module anyhow. Thoughts?

Comment (rafaelweingartner) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: It also works. I only think that we need to add something in the exception. I mean, a message saying what is not implemented (the preallocation type selected/sent by the user).

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java (hunk 1 above)
Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: check

Comment (nvazquez) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: Agree :+1:

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java [hunk 2]
```diff
@@ -399,4 +397,9 @@ public void resize(final QemuImgFile file, final long size, final boolean delta)
     public void resize(final QemuImgFile file, final long size) throws QemuImgException {
         this.resize(file, size, false);
     }
+
+    static class NotImplementedProvisioningTypeException extends RuntimeException {
+    }
 }
+
+;
```
Comment (nvazquez) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: Can this be removed please?

Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: of course, sorry

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java [hunk 3]
```diff
@@ -72,7 +71,7 @@ public static PreallocationType getPreallocationType(final Storage.ProvisioningT
             case FAT:
                 return PreallocationType.Full;
             default:
-                throw new NotImplementedException();
+                throw new NotImplementedProvisioningTypeException(String.format("type {} not defined as member-value of PreallocationType", provisioningType));
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: Is this a new type of `String.format` use? I am missing a `%s` there.

Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/qemu/QemuImg.java]: hm, i might be mixing python and java, will check.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Let's use org.apache.commons.lang.NotImplementedException which is already used by some storage component?

**Comment by nathanejohnson** [comment]
LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request merged**